### PR TITLE
Cleanup sockets when creating new connections

### PR DESCRIPTION
**Pull request by dmiddlecamp**
We noticed a bug where many thousands of simultaneous connections can be leaked.  This closes sockets and also prevents uncatchable exceptions from being thrown when many quick connection attempts are made in a short period of time.

### CODE REVIEW

**Comment by troy** [comment]
@kenperkins if we can help test and/or review this, I'm at your service

**Comment by kenperkins** [comment]
@dmiddlecamp can you expand on what you're seeing? I'd like to better understand.

**Comment by dmiddlecamp** [comment]
Sure sure!

We found an installation running this module that had spawned many thousands of connections to papertrail, which caused us to be rate limited.  Since the TLS session is created independently of the TLS session, it's possible for the module to attempt hundreds of simultaneous connections while it's waiting on the TLS session to be established.  These sockets aren't cleaned up, and the rate limiting wasn't enforced in this scenario.  This patch guarantees that the module will never open more than 1 socket at any given moment, and will cleanup the socket in the event that the TLS session doesn't succeed.

We were also seeing uncatchable crashes when many connection attempts were made quickly, so I've also added a fix for that and a unit test to confirm.

Thanks!
David

File: lib/winston-papertrail.js [hunk 1]
```diff
@@ -136,6 +136,40 @@ var Papertrail = exports.Papertrail = function (options) {
...
+        try {
+            if (self.socket) {
+                self.socket.destroy();
+
+                //this throws some uncatchable thing, eww.
+                //self.socket.end();
+                self.socket = null;
+            }
+        }
+        catch (e) { }
```
Comment (kenperkins) [inline lib/winston-papertrail.js]: So we're just catching here and swallowing the error. I'm wondering if we should emit it?

File: lib/winston-papertrail.js [hunk 2]
```diff
@@ -136,6 +136,40 @@ var Papertrail = exports.Papertrail = function (options) {
...
+            if (self.stream) {
+                //we're cleaning up, don't loop.
+                self.stream.removeListener('end', connectStream);
+                self.stream.removeListener('error', onErrored);
+
+                self.stream.destroy();
+                self.stream = null;
+            }
+        }
+        catch (e) { }
```
Comment (kenperkins) [inline lib/winston-papertrail.js]: Same as above.

File: lib/winston-papertrail.js [hunk 3]
```diff
@@ -265,9 +295,8 @@ Papertrail.prototype.log = function (level, msg, meta, callback) {
         meta = false;
     }
 
-    if  (meta && typeof meta === 'object' && (Object.keys(meta).length === 0)
-		&& (!util.isError(meta)))
-	{
+    if (meta && typeof meta === 'object' && (Object.keys(meta).length === 0)
+        && (!util.isError(meta))) {
```
Comment (kenperkins) [inline lib/winston-papertrail.js]: thanks for cleaning this up.

File: lib/winston-papertrail.js (hunk 1 above)
Comment (dmiddlecamp) [inline lib/winston-papertrail.js]: so the destroy here prevents the error from being thrown as opposed to close or end.  The try/catch does nothing to catch that socket error.

File: lib/winston-papertrail.js (hunk 2 above)
Comment (dmiddlecamp) [inline lib/winston-papertrail.js]: Here we're trying to ensure things are cleaned up properly to set the stage for the reconnection.  I don't think we should explode before we've had a chance to reconnect.  We're instantiating the library to talk to the logging provider, not detonate on cleanup.

File: lib/winston-papertrail.js (hunk 1 above)
Comment (dmiddlecamp) [inline lib/winston-papertrail.js]: Although we could certainly emit an error event on the class object, at least that would be catchable.  But it wouldn't be the aforementioned uncatchable socket error in TLS.

**Comment by dmiddlecamp** [comment]
Also BTW, thank you for taking the time to review, much appreciated! :)

File: lib/winston-papertrail.js (hunk 1 above)
Comment (kenperkins) [inline lib/winston-papertrail.js]: That's what I was wondering; is there value in `self.emit(e)`? Not being super familiar with when `socket.destroy()` throws I'm speculating.

File: lib/winston-papertrail.js (hunk 2 above)
Comment (kenperkins) [inline lib/winston-papertrail.js]: :)

Maybe it's better to discuss what stream or socket.destroy will throw and then we can be more accurate about the value of throwing here.

**Comment by kenperkins** [comment]
Did we decide whether or not we should emit those two exceptions?

I've typically biased to emitting so that if there's something weird going on, you bubble it up to a higher level where it might still be visible. I've learned that swallowing errors in your logging is generally bad :)

**Comment by dmiddlecamp** [comment]
I'm okay with emitting those if you feel strongly about it, but I only try to emit errors for things where something should / can be done.  In a scorched earth / cleanup scenario I'm not sure what recovery steps we'd take out of band between disposing sockets and re-establishing them?

**Comment by kenperkins** [comment]
I think the only thing that could be done is being able to log something out to the console for investigation purposes. And given that I don't even know what this will throw, lets not bother for now. 

If we see evidence to the contrary, we can revisit.

**Comment by matteocontrini** [comment]
I tend to agree with @dmiddlecamp, I wouldn't know what to do knowing that winston-papertrail had an internal error. Would I log that the logging library had an error? :P

**Comment by kenperkins** [comment]
So be it :)

**Comment by dcollinsf5** [comment]
Hi guys!
I believe that this issue caused an outage for our production app this last week during a papertrail outage.  Could you guys shed some light on this pull req; does it resolve the many-connections issue, and if so why isn't it merged?

**Comment by troy** [comment]
@dmiddlecamp could I trouble you to pull master and merge it? We're merging this and an unrelated PR into master, then letting folks use it for a few weeks before making a release.

**Comment by dmiddlecamp** [comment]
Sure thing, I'll take a look

**Comment by dmiddlecamp** [comment]
Looks like it was just a package.json conflict?

**Comment by troy** [comment]
If you're reading this message, we'd love if you can pull the current master of winston-papertrail and run it. We'll wait at least 2 weeks to see whether any issues come up from wider use of this and one other change. If you encounter problems with master which don't occur in the current release, please open a new GitHub Issue with the transcript and reference this PR.

**Comment by troy** [comment]
This is live on NPM in version `1.0.3`.

**Pull request merged**